libpod/container.go

```diff
@@ -1,69 +1,126 @@
-package ctr
+package libpod
 
 import (
 	"fmt"
+	"sync"
 
 	"github.com/containers/storage"
+	"github.com/docker/docker/pkg/stringid"
+	spec "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/ulule/deepcopier"
 )
 
 var (
 	// ErrNotImplemented indicates that functionality is not yet implemented
-	ErrNotImplemented = fmt.Errorf("NOT IMPLEMENTED")
+	errNotImplemented = fmt.Errorf("NOT IMPLEMENTED")
 )
 
 // Container is a single OCI container
 type Container struct {
-	// TODO populate
+	id   string
+	name string
+
+	spec *spec.Spec
+	pod  *Pod
+
+	valid bool
+	lock  sync.RWMutex
+}
+
+// ID returns the container's ID
+func (c *Container) ID() string {
+	// No locking needed, ID will never mutate after a container is created
+	return c.id
+}
+
+// Name returns the container's name
+func (c *Container) Name() string {
+	// Name can potentially be changed while a container is running
+	// So lock access to it
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	return c.name
+}
+
+// Spec returns the container's OCI runtime spec
+func (c *Container) Spec() *spec.Spec {
+	// The spec can potentially be altered when storage is configured and to
+	// add annotations at container create time
+	// As such, access to it is locked
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	spec := new(spec.Spec)
+	deepcopier.Copy(c.spec).To(spec)
+
+	return spec
+}
+
+// Make a new container
+func newContainer(rspec *spec.Spec) (*Container, error) {
+	if rspec == nil {
+		return nil, fmt.Errorf("must provide a valid spec to construct container")
+	}
+
+	ctr := new(Container)
+	ctr.id = stringid.GenerateNonCryptoID()
+	ctr.name = ctr.id // TODO generate unique human-readable names
```

Comment:
Should we steal the moby code for this?

Reply:
I'd be fine with this, but this PR is already huge, so I'd prefer to leave the extra vendoring to later.

Reply:
SGTM